File: frontend/src/api/applications.ts
import { del, get, patch, post } from "./client";
import type { ApplicationCreate, ApplicationRead, ApplicationUpdate, BulkUpdate } from "../types";

interface ListParams {
  status?: string;
  company_id?: number;
  include_archived?: boolean;
  ever_status?: string[];
  ever_status_match_all?: boolean;
}

export function listApplications(params: ListParams = {}): Promise<ApplicationRead[]> {
  const qs = new URLSearchParams();
  if (params.status) qs.set("status", params.status);
  if (params.company_id) qs.set("company_id", String(params.company_id));
  if (params.include_archived) qs.set("include_archived", "true");
  params.ever_status?.forEach((s) => qs.append("ever_status", s));
  if (params.ever_status_match_all) qs.set("ever_status_match_all", "true");
  const query = qs.toString();
  return get<ApplicationRead[]>(`/applications${query ? `?${query}` : ""}`);
}

export const createApplication = (data: ApplicationCreate) =>
  post<ApplicationRead>("/applications", data);

export const getApplication = (id: number) =>
  get<ApplicationRead>(`/applications/${id}`);

export const updateApplication = (id: number, data: ApplicationUpdate) =>
  patch<ApplicationRead>(`/applications/${id}`, data);

export const deleteApplication = (id: number) =>
  del(`/applications/${id}`);

export const archiveApplication = (id: number) =>
  post<ApplicationRead>(`/applications/${id}/archive`);

export const unarchiveApplication = (id: number) =>
  post<ApplicationRead>(`/applications/${id}/unarchive`);

export const bulkUpdate = (data: BulkUpdate) =>
  patch<ApplicationRead[]>("/applications/bulk", data);

export const bulkDelete = (ids: number[]) =>
  post<void>("/applications/bulk-delete", { ids });

File: backend/app/api/routes/applications.py
from typing import Optional

from fastapi import APIRouter, Query

from app.api.deps import SessionDep
from app.schemas.application import (
    ApplicationCreate,
    ApplicationRead,
    ApplicationUpdate,
    BulkDelete,
    BulkUpdate,
)
from app.services import application_service

router = APIRouter(prefix="/applications", tags=["applications"])


@router.get("", response_model=list[ApplicationRead])
def list_applications(
    session: SessionDep,
    status: Optional[str] = None,
    company_id: Optional[int] = None,
    include_archived: bool = False,
    ever_status: Optional[list[str]] = Query(None),
    ever_status_match_all: bool = False,
):
    return application_service.list_all(
        session,
        include_archived=include_archived,
        company_id=company_id,
        status=status,
        ever_status=ever_status,
        ever_status_match_all=ever_status_match_all,
    )


@router.post("", response_model=ApplicationRead, status_code=201)
def create_application(data: ApplicationCreate, session: SessionDep):
    return application_service.create(session, data)


@router.get("/{app_id}", response_model=ApplicationRead)
def get_application(app_id: int, session: SessionDep):
    return application_service.get(session, app_id)


@router.patch("/bulk", response_model=list[ApplicationRead])
def bulk_update_applications(data: BulkUpdate, session: SessionDep):
    return application_service.bulk_update(session, data)


@router.post("/bulk-delete", status_code=204)
def bulk_delete_applications(data: BulkDelete, session: SessionDep):
    application_service.bulk_delete(session, data)


@router.patch("/{app_id}", response_model=ApplicationRead)
def update_application(app_id: int, data: ApplicationUpdate, session: SessionDep):
    return application_service.update(session, app_id, data)


@router.delete("/{app_id}", status_code=204)
def delete_application(app_id: int, session: SessionDep):
    application_service.delete(session, app_id)


@router.post("/{app_id}/archive", response_model=ApplicationRead)
def archive_application(app_id: int, session: SessionDep):
    return application_service.archive(session, app_id)


@router.post("/{app_id}/unarchive", response_model=ApplicationRead)
def unarchive_application(app_id: int, session: SessionDep):
    return application_service.unarchive(session, app_id)
